### PR TITLE
Add Para threads to MMR Root

### PR DESCRIPTION
**Pull request by alistair-singh**
Fixes related to: https://github.com/paritytech/polkadot-sdk/pull/4751

We may need to time this to make sure we only include this when the PR release is live.

### CODE REVIEW

File: relayer/relays/parachain/beefy-listener.go [hunk 1]
```diff
@@ -258,8 +262,10 @@ func (li *BeefyListener) generateProof(ctx context.Context, input *ProofInput, h
 	// Generate a merkle proof for the parachain head with input ParaId
 	// and verify with merkle root hash of all parachain heads
 	// Polkadot uses the following code to generate merkle root from parachain headers:
-	// https://github.com/paritytech/polkadot/blob/2eb7672905d99971fc11ad7ff4d57e68967401d2/runtime/rococo/src/lib.rs#L706-L709
-	merkleProofData, err := CreateParachainMerkleProof(input.ParaHeads, input.ParaID)
+	// https://github.com/paritytech/polkadot-sdk/blob/d66dee3c3da836bcf41a12ca4e1191faee0b6a5b/polkadot/runtime/westend/src/lib.rs#L453-L460
+	// Truncate the ParaHeads to the 1024
+	// https://github.com/paritytech/polkadot-sdk/blob/d66dee3c3da836bcf41a12ca4e1191faee0b6a5b/polkadot/runtime/parachains/src/paras/mod.rs#L1305-L1311
+	merkleProofData, err := CreateParachainMerkleProof(input.ParaHeads[:MaxParaHeads], input.ParaID)
```
Comment (yrong) [inline relayer/relays/parachain/beefy-listener.go:268]: So are we always generate proof against a static length array(i.e. 1024)? Or could we use `min(len(paraHeads),1024)`?

Comment (alistair-singh) [reply relayer/relays/parachain/beefy-listener.go]: Good catch! Added in [1d4c23a](https://github.com/Snowfork/snowbridge/pull/1288/commits/1d4c23a45c97563a828951ac35a06026e6b11f22).

**Comment by alistair-singh** [comment]
Tested successfully in rococo and it has fixed the issue.

**Review by yrong — APPROVED** [review]
+1

File: relayer/chain/relaychain/connection.go [hunk 2]
```diff
@@ -292,7 +218,8 @@ func (co *Connection) fetchParaHeads(blockHash types.Hash) (map[uint32]ParaHead,
 		return nil, err
 	}
 
-	heads := make(map[uint32]ParaHead)
+	const numParas = 16
```
Comment (claravanstaden) [inline relayer/chain/relaychain/connection.go:221]: Where does this number come from?

Comment (yrong) [reply relayer/chain/relaychain/connection.go:221]: Just initial compacity of the array which can be extended as required.

File: relayer/chain/relaychain/connection.go [hunk 3]
```diff
@@ -221,58 +191,15 @@ func (co *Connection) FetchMMRLeafCount(relayBlockhash types.Hash) (uint64, erro
 	return mmrLeafCount, nil
 }
 
-func (co *Connection) fetchKeys(keyPrefix []byte, blockHash types.Hash) ([]types.StorageKey, error) {
```
Comment (yrong) [inline relayer/chain/relaychain/connection.go:224]: Seems just move position of this function, there is no internal change?

If so would suggest to leave it untouched to make diff a bit easier.

Comment (alistair-singh) [reply relayer/chain/relaychain/connection.go]: addressed in https://github.com/Snowfork/snowbridge/pull/1288/commits/f6b090c7bcedde5f79d34eeacc4de80d21dc0614

**Review by vgeddes — APPROVED** [review]

**Pull request merged**